src/batch.rs

```diff
@@ -0,0 +1,183 @@
+    public_keys: &[PublicKey],
+    messages: &[&[u8]],
+    mut rng: impl CryptoRng + RngCore,
+) -> Result<(), SignatureError> {
+    assert!(
+        signatures.len() == public_keys.len(),
+        "We should have the same number of signatures than public keys"
+    );
+    assert!(
+        messages.len() == public_keys.len(),
+        "We should have the same number of messages than public keys"
+    );
+
+    let (scalars, hashes) =
+        generate_batch_coefficients(signatures, public_keys, messages, &mut rng);
+
+    verify_prepared_batch(scalars, hashes, signatures, public_keys)
+}
+
+/// Prepares a batch verification of Schnorr signatures
+/// It computes the random challenges for each signature and the random scalars to
+/// scale the public keys.
+#[allow(non_snake_case)]
+fn generate_batch_coefficients(
+    signatures: &[Signature],
+    public_keys: &[PublicKey],
+    messages: &[&[u8]],
+    mut rng: impl CryptoRng + RngCore,
+) -> (Vec<Scalar>, Vec<Scalar>) {
+    let hashes: Vec<Scalar> = signatures
+        .iter()
+        .zip(public_keys)
+        .zip(messages)
+        .map(|((s, k), m)| {
+            let x_felt = Fp6::from_bytes(&s.x.0[0..48].try_into().unwrap()).unwrap();
+            let h = hash_message(&x_felt, k, m);
+            let h_bits = h.as_bits::<Lsb0>();
+
+            Scalar::from_bits_vartime(h_bits)
+        })
+        .collect();
+
+    let scalars: Vec<Scalar> = signatures
+        .iter()
+        .map(|_| Scalar::random(&mut rng))
+        .collect();
+
+    (scalars, hashes)
+}
+
+/// Verifies a batch with a set of random scalars and hash outputs
+fn verify_prepared_batch(
+    scalars: Vec<Scalar>,
+    mut hashes: Vec<Scalar>,
+    signatures: &[Signature],
+    public_keys: &[PublicKey],
+) -> Result<(), SignatureError> {
+    // Compute the linear combination of the random scalars with the
+    // signatures. This is used to multiply the curve basepoint.
+    let lin_comb: Scalar = signatures
+        .iter()
+        .map(|sig| sig.e)
+        .zip(scalars.iter())
+        .map(|(e, s)| s * e)
+        .sum();
+    let scaled_basepoint: AffinePoint = BASEPOINT_TABLE
+        .multiply_vartime(&lin_comb.to_bytes())
+        .into();
+
+    let mut points: Vec<AffinePoint> = signatures
+        .iter()
+        .map(|sig| AffinePoint::from_compressed(&sig.x).unwrap())
+        .collect();
+    points.append(&mut public_keys.iter().map(|k| k.0.neg()).collect());
+
+    // Multiply each hash by the random value
+    for (h, s) in hashes.iter_mut().zip(scalars.iter()) {
+        *h *= s;
+    }
+
+    // Convert scalars and hash outputs to byte slices
+    let scalar_bytes: Vec<[u8; 32]> = scalars
+        .into_iter()
+        .chain(hashes.into_iter())
+        .map(|s| s.to_bytes())
+        .collect();
+
+    // Compute the multi-scalar multiplication
+    // This differs from a single Schnorr verification, as we check
+    // Σ s[i].signatures[i].x - Σ s[i]h[i].p[i] = Σ s[i].G
```

Comment:
Shouldn't the RHS be (Σ s[i].signatures[i].e).G ?

Reply:
Right there's a typo, good catch!